File: php/src/Phc/PhcParser.php
<?php

declare(strict_types=1);

namespace ArgonGuard\Passwords\Phc;

use ArgonGuard\Passwords\MalformedHashException;
use ArgonGuard\Passwords\ReasonCodes;

/**
 * 嚴格 PHC parser／encoder（SPEC §2、§4 S1–S4；baseline §4 嚴格文法澄清）。
 *
 * 文法澄清（跨語言必須 bit-identical，由 .NET baseline 釘死）：
 * 數字欄位僅允許 [0-9]、不允許正負號、不允許前導零（值本身為 0 除外）、位數上限 15（防溢位）；
 * base64 採 RFC 4648 §4 標準字元集、無 padding，且必須 canonical
 * （decode 後 re-encode 必須等於原字串，封死 trailing-bit 可鍛性）；長度 mod 4 == 1 → 非法。
 *
 * @internal 不屬於公開 API。
 */
final class PhcParser
{
    private const BASE64_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    private const DIGITS = '0123456789';

    private function __construct()
    {
    }

    /**
     * 抽出 PHC 演算法 token（dispatch 前置判斷；baseline §1）：字串為 "$<token>$…" 且 token 合法
     * （小寫英數與 '-'）時回傳 token，否則 null。token != "argon2id" → dispatch 走
     * UnsupportedAlgorithm 路徑，不套用 argon2 嚴格文法。
     */
    public static function tryGetAlgorithm(string $encoded): ?string
    {
        if (strlen($encoded) < 3 || $encoded[0] !== '$') {
            return null;
        }
        $end = strpos($encoded, '$', 1);
        if ($end === false || $end <= 1) {
            return null;
        }
        $token = substr($encoded, 1, $end - 1);

        return self::isLowerAlnumDash($token) ? $token : null;
    }

    /** 嚴格解析。失敗拋 MalformedHashException（reason 依 SPEC）。長度預檢（>512）由呼叫端負責。 */
    public static function parse(string $encoded): PhcHash
    {
        if ($encoded === '' || $encoded[0] !== '$') {
            throw new MalformedHashException(ReasonCodes::NOT_PHC);
        }

        $parts = explode('$', $encoded);
        // ["", alg, "v=19", params, salt, tag]（有 v）或 ["", alg, params, salt, tag]（缺 v → 政策層 missing_version）
        $count = count($parts);
        if ($count !== 5 && $count !== 6) {
            throw new MalformedHashException(ReasonCodes::NOT_PHC);
        }

        $algorithm = $parts[1];
        if ($algorithm === '' || !self::isLowerAlnumDash($algorithm)) {
            throw new MalformedHashException(ReasonCodes::NOT_PHC);
        }

        $version = null;
        $paramsIndex = 2;
        if ($count === 6) {
            // v 段有出現時必須是第 3 段且格式 v=<number>（baseline §4）
            if (!str_starts_with($parts[2], 'v=')) {
                throw new MalformedHashException(ReasonCodes::NOT_PHC);
            }
            $version = self::parseNumber(substr($parts[2], 2));
            $paramsIndex = 3;
        }

        [$m, $t, $p, $hasKeyid, $hasData] = self::parseParams($parts[$paramsIndex]);
        $salt = self::decodeCanonicalBase64($parts[$paramsIndex + 1]);
        $tag = self::decodeCanonicalBase64($parts[$paramsIndex + 2]);

        return new PhcHash($algorithm, $version, $m, $t, $p, $salt, $tag, $hasKeyid, $hasData);
    }

    /** 產生端 encoder（SPEC §2 G1–G8）。salt/tag 為 raw bytes。 */
    public static function encode(int $m, int $t, int $p, string $salt, string $tag): string
    {
        return sprintf(
            '$argon2id$v=19$m=%d,t=%d,p=%d$%s$%s',
            $m,
            $t,
            $p,
            self::encodeBase64NoPad($salt),
            self::encodeBase64NoPad($tag),
        );
    }

    /** @return array{0:int,1:int,2:int,3:bool,4:bool} [m, t, p, hasKeyid, hasData] */
    private static function parseParams(string $paramSegment): array
    {
        $tokens = explode(',', $paramSegment);
        if (count($tokens) < 3) {
            throw new MalformedHashException(
                self::isPermutedMtp($tokens) ? ReasonCodes::PARAMS_OUT_OF_ORDER : ReasonCodes::NOT_PHC,
            );
        }

        // 前三個 token 必須依序為 m=、t=、p=（SPEC S1）
        if (!(str_starts_with($tokens[0], 'm=')
            && str_starts_with($tokens[1], 't=')
            && str_starts_with($tokens[2], 'p='))) {
            throw new MalformedHashException(
                self::isPermutedMtp($tokens) ? ReasonCodes::PARAMS_OUT_OF_ORDER : ReasonCodes::NOT_PHC,
            );
        }

        $m = self::parseNumber(substr($tokens[0], 2));
        $t = self::parseNumber(substr($tokens[1], 2));
        $p = self::parseNumber(substr($tokens[2], 2));

        $hasKeyid = false;
        $hasData = false;
        $tokenCount = count($tokens);
        for ($i = 3; $i < $tokenCount; $i++) {
            // 第 4 個起僅允許 keyid=／data= 前綴 token（→ 政策層拒絕），其他 → not_phc（baseline §4）
            if (str_starts_with($tokens[$i], 'keyid=')) {
                $hasKeyid = true;
            } elseif (str_starts_with($tokens[$i], 'data=')) {
                $hasData = true;
            } else {
                throw new MalformedHashException(ReasonCodes::NOT_PHC);
            }
        }

        return [$m, $t, $p, $hasKeyid, $hasData];
    }

    /**
     * 前三 token 是否為 m/t/p 的重排（區分 params_out_of_order 與 not_phc）。
     *
     * @param list<string> $tokens
     */
    private static function isPermutedMtp(array $tokens): bool
    {
        if (count($tokens) < 3) {
            return false;
        }
        $seen = 0;
        for ($i = 0; $i < 3; $i++) {
            if (str_starts_with($tokens[$i], 'm=')) {
                $seen |= 1;
            } elseif (str_starts_with($tokens[$i], 't=')) {
                $seen |= 2;
            } elseif (str_starts_with($tokens[$i], 'p=')) {
                $seen |= 4;
            } else {
                return false;
            }
        }

        return $seen === 7;
    }

    /** 嚴格數字：僅 [0-9]、無正負號、無前導零（單獨 "0" 除外）、位數 ≤15（PHP int 64 位，15 位十進位不溢位）。 */
    private static function parseNumber(string $digits): int
    {
        $len = strlen($digits);
        if ($len === 0 || $len > 15) {
            throw new MalformedHashException(ReasonCodes::NOT_PHC);
        }
        if ($len > 1 && $digits[0] === '0') {
            throw new MalformedHashException(ReasonCodes::NOT_PHC); // 禁前導零（嚴格文法）
        }
        if (strspn($digits, self::DIGITS) !== $len) {
            throw new MalformedHashException(ReasonCodes::NOT_PHC);
        }

        return (int) $digits;
    }

    /** 回傳 raw bytes。含 '='（padding）、base64url、其他字元、非 canonical 一律 bad_base64。 */
    private static function decodeCanonicalBase64(string $s): string
    {
        $len = strlen($s);
        if ($len === 0 || strspn($s, self::BASE64_ALPHABET) !== $len) {
            throw new MalformedHashException(ReasonCodes::BAD_BASE64);
        }
        $rem = $len % 4;
        if ($rem === 1) {
            throw new MalformedHashException(ReasonCodes::BAD_BASE64);
        }
        $decoded = base64_decode($s . str_repeat('=', (4 - $rem) % 4), true);
        if ($decoded === false) {
            throw new MalformedHashException(ReasonCodes::BAD_BASE64);
        }
        // canonical 檢查：re-encode 必須還原原字串（封死 trailing-bit 可鍛性）
        if (self::encodeBase64NoPad($decoded) !== $s) {
            throw new MalformedHashException(ReasonCodes::BAD_BASE64);
        }

        return $decoded;
    }

    private static function encodeBase64NoPad(string $data): string
    {
        return rtrim(base64_encode($data), '=');
    }

    private static function isLowerAlnumDash(string $s): bool
    {
        return strspn($s, 'abcdefghijklmnopqrstuvwxyz0123456789-') === strlen($s);
    }
}

File: dotnet/src/ArgonGuard.Passwords/Phc/PhcParser.cs
using System;
using System.Text;

namespace ArgonGuard.Passwords.Phc;

/// <summary>
/// 嚴格 PHC parser／encoder（SPEC §2、§4 S1–S4）。
/// 文法澄清（跨語言必須一致，由單元測試與 baseline 文件釘死）：
/// 數字欄位僅允許 [0-9]、不允許前導零（值本身為 0 除外）、位數上限 15（防溢位，超限視為超天花板方向於政策層處理）；
/// base64 採 RFC 4648 §4 標準字元集、無 padding，且必須 canonical（decode 後 re-encode 必須等於原字串）。
/// </summary>
internal static class PhcParser
{
    private const string Argon2idPrefix = "argon2id";

    /// <summary>
    /// 抽出 PHC 演算法 token（dispatch 前置判斷）：字串為 "$&lt;token&gt;$…" 且 token 合法（小寫英數與 '-'）
    /// 時回傳 token，否則 null。token != "argon2id" → dispatch 走 UnsupportedAlgorithm 路徑，
    /// 不套用 argon2 嚴格文法（跨語言一致，baseline 文件釘死）。
    /// </summary>
    public static string? TryGetAlgorithm(string encoded)
    {
        if (encoded.Length < 3 || encoded[0] != '$') return null;
        int end = encoded.IndexOf('$', 1);
        if (end <= 1) return null;
        var token = encoded.Substring(1, end - 1);
        return IsLowerAlnumDash(token) ? token : null;
    }

    /// <summary>嚴格解析。失敗拋 MalformedHashException（reason 依 SPEC）。長度預檢（&gt;512）由呼叫端負責。</summary>
    public static PhcHash Parse(string encoded)
    {
        if (string.IsNullOrEmpty(encoded) || encoded[0] != '$')
            throw new MalformedHashException(ReasonCodes.NotPhc);

        var parts = encoded.Split('$');
        // ["", alg, "v=19", params, salt, tag]（有 v）或 ["", alg, params, salt, tag]（缺 v → 政策層 missing_version）
        if (parts.Length is not (5 or 6))
            throw new MalformedHashException(ReasonCodes.NotPhc);

        var algorithm = parts[1];
        if (algorithm.Length == 0 || !IsLowerAlnumDash(algorithm))
            throw new MalformedHashException(ReasonCodes.NotPhc);

        int? version = null;
        int paramsIndex = 2;
        if (parts.Length == 6)
        {
            if (!parts[2].StartsWith("v=", StringComparison.Ordinal))
                throw new MalformedHashException(ReasonCodes.NotPhc);
            version = (int)ParseNumber(parts[2].Substring(2));
            paramsIndex = 3;
        }

        var (m, t, p, hasKeyid, hasData) = ParseParams(parts[paramsIndex]);
        var salt = DecodeCanonicalBase64(parts[paramsIndex + 1]);
        var tag = DecodeCanonicalBase64(parts[paramsIndex + 2]);
        return new PhcHash(algorithm, version, m, t, p, salt, tag, hasKeyid, hasData);
    }

    /// <summary>產生端 encoder（SPEC §2 G1–G8）。</summary>
    public static string Encode(long m, long t, long p, byte[] salt, byte[] tag)
        => $"$argon2id$v=19$m={m},t={t},p={p}${EncodeBase64NoPad(salt)}${EncodeBase64NoPad(tag)}";

    private static (long M, long T, long P, bool HasKeyid, bool HasData) ParseParams(string paramSegment)
    {
        var tokens = paramSegment.Split(',');
        if (tokens.Length < 3)
            throw new MalformedHashException(IsPermutedMtp(tokens) ? ReasonCodes.ParamsOutOfOrder : ReasonCodes.NotPhc);

        // 前三個 token 必須依序為 m=、t=、p=（SPEC S1）
        if (!(tokens[0].StartsWith("m=", StringComparison.Ordinal)
              && tokens[1].StartsWith("t=", StringComparison.Ordinal)
              && tokens[2].StartsWith("p=", StringComparison.Ordinal)))
        {
            throw new MalformedHashException(IsPermutedMtp(tokens) ? ReasonCodes.ParamsOutOfOrder : ReasonCodes.NotPhc);
        }

        long m = ParseNumber(tokens[0].Substring(2));
        long t = ParseNumber(tokens[1].Substring(2));
        long p = ParseNumber(tokens[2].Substring(2));

        bool hasKeyid = false, hasData = false;
        for (int i = 3; i < tokens.Length; i++)
        {
            if (tokens[i].StartsWith("keyid=", StringComparison.Ordinal)) hasKeyid = true;
            else if (tokens[i].StartsWith("data=", StringComparison.Ordinal)) hasData = true;
            else throw new MalformedHashException(ReasonCodes.NotPhc);
        }
        return (m, t, p, hasKeyid, hasData);
    }

    /// <summary>前三 token 是否為 m/t/p 的重排（區分 params_out_of_order 與 not_phc）。</summary>
    private static bool IsPermutedMtp(string[] tokens)
    {
        if (tokens.Length < 3) return false;
        int seen = 0;
        for (int i = 0; i < 3; i++)
        {
            if (tokens[i].StartsWith("m=", StringComparison.Ordinal)) seen |= 1;
            else if (tokens[i].StartsWith("t=", StringComparison.Ordinal)) seen |= 2;
            else if (tokens[i].StartsWith("p=", StringComparison.Ordinal)) seen |= 4;
            else return false;
        }
        return seen == 7;
    }

    private static long ParseNumber(string digits)
    {
        if (digits.Length is 0 or > 15)
            throw new MalformedHashException(ReasonCodes.NotPhc);
        if (digits.Length > 1 && digits[0] == '0')
            throw new MalformedHashException(ReasonCodes.NotPhc); // 禁前導零（嚴格文法）
        long value = 0;
        foreach (var c in digits)
        {
            if (c is < '0' or > '9')
                throw new MalformedHashException(ReasonCodes.NotPhc);
            value = value * 10 + (c - '0');
        }
        return value;
    }

    private const string Base64Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    private static byte[] DecodeCanonicalBase64(string s)
    {
        if (s.Length == 0)
            throw new MalformedHashException(ReasonCodes.BadBase64);
        foreach (var c in s)
        {
            if (Base64Alphabet.IndexOf(c) < 0)
                throw new MalformedHashException(ReasonCodes.BadBase64); // 含 '='（padding）、base64url、其他字元一律拒絕
        }
        int rem = s.Length % 4;
        if (rem == 1)
            throw new MalformedHashException(ReasonCodes.BadBase64);
        byte[] decoded;
        try
        {
            decoded = Convert.FromBase64String(s + new string('=', (4 - rem) % 4));
        }
        catch (FormatException)
        {
            throw new MalformedHashException(ReasonCodes.BadBase64);
        }
        // canonical 檢查：re-encode 必須還原原字串（封死 trailing-bit 可鍛性）
        if (!string.Equals(EncodeBase64NoPad(decoded), s, StringComparison.Ordinal))
            throw new MalformedHashException(ReasonCodes.BadBase64);
        return decoded;
    }

    private static string EncodeBase64NoPad(byte[] data) => Convert.ToBase64String(data).TrimEnd('=');

    private static bool IsLowerAlnumDash(string s)
    {
        foreach (var c in s)
        {
            if (c is not ((>= 'a' and <= 'z') or (>= '0' and <= '9') or '-'))
                return false;
        }
        return true;
    }
}
